File: 1912-number-of-restricted-paths-from-first-to-last-node/1912-number-of-restricted-paths-from-first-to-last-node.py
from heapq import heappop, heappush
from typing import List

MOD = 10**9 + 7
LL_INF = 10**18
class Solution:
    def __init__(self):
        self.n = 0

    def dijkstra(self, distance: List[int], src: int, adj: List[List[List[int]]]):
        pq = []
        heappush(pq, (0, src))
        distance[src] = 0

        while pq:
            dis, node = heappop(pq)
            for neighbor, wt in adj[node]:
                if dis + wt < distance[neighbor]:
                    distance[neighbor] = dis + wt
                    heappush(pq, (distance[neighbor], neighbor))
    
    def memoization(self, node: int, adj: List[List[List[int]]], distance: List[int], dp: List[int]) -> int:
        if node == self.n:
            return 1
        if dp[node] != -1:
            return dp[node]
        
        ans = 0
        for neighbor, _ in adj[node]:
            if distance[neighbor] < distance[node]:
                ans = (ans + self.memoization(neighbor, adj, distance, dp)) % MOD
        
        dp[node] = ans
        return dp[node]
    def countRestrictedPaths(self, N: int, edges: List[List[int]]) -> int:
        self.n = N
        adj = [[] for _ in range(N + 1)]
        for u, v, wt in edges:
            adj[u].append([v, wt])
            adj[v].append([u, wt])
        
        distance = [LL_INF] * (N + 1)
        dp = [-1] * (N + 1)
        self.dijkstra(distance, N, adj)
        return self.memoization(1, adj, distance, dp)

File: 1912-number-of-restricted-paths-from-first-to-last-node/1912-number-of-restricted-paths-from-first-to-last-node.cpp
const int mod = 1e9+7;
using ll = long long int;
class Solution {
public:
    int n;
    void dijkstra(vector<ll> &distance, ll src, vector<vector<ll>> adj[]){
        priority_queue<pair<ll,ll>, vector<pair<ll,ll>>, greater<pair<ll,ll>>> pq;
        pq.push({0,src}); distance[src] = 0;
        while (!pq.empty()){
            auto topNode = pq.top(); pq.pop();
            ll dis = topNode.first, node = topNode.second;
            for (auto it : adj[node]){
                ll neighbor = it[0], wt = it[1];
                if (dis + wt < distance[neighbor]){
                    distance[neighbor] = dis + wt;
                    pq.push({dis + wt,neighbor});
                } 
            } 
        }
        return ;
    }
    ll memoization(ll node, vector<vector<ll>> adj[], vector<ll> &distance, vector<ll>& dp){
        if (node == n) return 1;
        if (dp[node]!=-1) return dp[node];
        ll ans = 0;
        for (auto it : adj[node]){
            if (distance[it[0]] < distance[node]){
                ans = (ans + memoization(it[0], adj, distance, dp))%mod;
            }
        }
        return dp[node] = ans;
    }
    ll countRestrictedPaths(int N, vector<vector<int>>& edges) {
        n = N;
        vector<vector<ll>> adj[n+1];
        for (auto it : edges){
            ll u = it[0], v = it[1], wt = it[2];
            adj[u].push_back({v,wt});
            adj[v].push_back({u,wt});
        }
        vector<ll> distance(n+1, 1e18), dp(n+1, -1);
        dijkstra(distance, n, adj);
        return memoization(1, adj, distance, dp);

    }
};